File: src/main/scala/com/outsmart/measurement/TimedValue.scala
package com.outsmart.measurement

/**
 * @author Vadim Bobrov
*/
class TimedValue(val timestamp : Long, val value : Double) extends Ordered[TimedValue] {
  	override def toString = "ts: " + timestamp + " value: " + value
  	override def compare(that : TimedValue) : Int = this.timestamp.compareTo(that.timestamp)

  	override def equals(that : Any)  = {
		that.isInstanceOf[TimedValue] && that.asInstanceOf[TimedValue].timestamp == this.timestamp
  	}

  	// define addition as adding values
  	def +(that: TimedValue): TimedValue = new TimedValue(timestamp, value + that.value)

	// by default use energy as value
	implicit def measuredValueToTimedValue(x: MeasuredValue) = new TimedValue(x.timestamp, x.energy)
}

File: src/main/java/com/outsmart/measurement/TimedValue.java
package com.outsmart.measurement;

/**
 * @author Vadim Bobrov
 */
public class TimedValue implements Comparable<TimedValue> {

	private final long timestamp;
	private final double value;

	public TimedValue(long timestamp, double value) {
		this.timestamp = timestamp;
		this.value = value;
	}

	public long timestamp() {
		return timestamp;
	}

	public double value() {
		return value;
	}

	public TimedValue add(TimedValue that) {
		return new TimedValue(this.timestamp(), this.value() + that.value());
	}

	@Override
	public String toString() {
		return "ts: " + timestamp + " value: " + value;
	}


	@Override
	public int compareTo(TimedValue o) {
		return this.timestamp - o.timestamp < 0 ? -1 : (this.timestamp - o.timestamp > 0 ? 1 : 0 );
	}


	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		TimedValue that = (TimedValue) o;

		if (timestamp != that.timestamp) return false;

		return true;
	}

	@Override
	public int hashCode() {
		return (int) (timestamp ^ (timestamp >>> 32));
	}

}
